contracts/RelayerHub.sol

```diff
@@ -8,88 +8,192 @@ import "./System.sol";
+
+            require(value.length == 20, "length of manager address mismatch");
+            address newManager = BytesToTypes.bytesToAddress(20, value);
+            addManagerByGov(newManager);
+
+        } else if (Memory.compareStrings(key, "removeManager")) {
+
+            require(value.length == 20, "length of manager address mismatch");
+            address payable managerAddress = payable(BytesToTypes.bytesToAddress(20, value));
+            removeManager(managerAddress);
+
+        } else {
+            require(false, "unknown param");
+        }
+        emit paramChange(key, value);
+    }
+
+    function removeManagerByHimself() external {
+        // here the manager removes himself
+        removeManager(payable(msg.sender));
+    }
+
+    function removeManager(address payable managerAddress) internal {
+        // check if the manager address already exists
+        require(relayManagersExistMap[managerAddress], "manager doesn't exist");
+
+        address relayerAddress = managerToRelayer[managerAddress];
+
+        manager memory m = managers[managerAddress];
+
+        delete (relayManagersExistMap[managerAddress]);
+        delete (managerToRelayer[managerAddress]);
+
+        address payable systemPayable = payable(address(uint160(SYSTEM_REWARD_ADDR)));
+        systemPayable.transfer(m.dues);
+
+        delete(managers[managerAddress]);
+
+
+        // emit success event
+        emit removeManagerEvent(managerAddress);
+        if (relayerAddress != address(0)) {
+            delete (currentRelayers[relayerAddress]);
+            emit updateRelayerEvent(relayerAddress, address(0));
+        }
+    }
+
+    function addManagerByGov(address managerToBeAdded) internal {
+        require(!relayManagersExistMap[managerToBeAdded], "manager already exists");
+        require(!isContract(managerToBeAdded), "contract is not allowed to be a manager");
+
+        relayManagersExistMap[managerToBeAdded] = true;
+
+        emit addManagerByGovEvent(managerToBeAdded);
+    }
+
+    // updateRelayer() can be used to add relayer for the first time, update it in future and remove it
+    // in case of removal we can simply update it to a non-existing account
+    function updateRelayer(address relayerToBeAdded) public onlyRegisteredManager {
+        if (relayerToBeAdded != address(0)) {
+            require(!currentRelayers[relayerToBeAdded], "relayer already exists");
+        }
+        require(!isContract(relayerToBeAdded), "contract is not allowed to be a relayer");
+
+        address oldRelayer = managerToRelayer[msg.sender];
+        delete currentRelayers[oldRelayer];
+
+        managerToRelayer[msg.sender] = relayerToBeAdded;
+        currentRelayers[relayerToBeAdded] = true;
+
+        emit updateRelayerEvent(oldRelayer, relayerToBeAdded);
+    }
+
+    function registerManagerAddRelayer(address r) external payable onlyNonRegisteredManager {
+        // register manager
+        managers[msg.sender] = manager(dues, true);
```

Comment:
A potential bug here:  You did not check the `msg.value == dues`, so that the msg value can be zero, but RelayerHub will transfer dues to system reward contract, where does the funds come?

Reply:
I will suggest no `dues` at all to make it simple.